htlcswitch/link.go

```diff
@@ -94,6 +97,56 @@ type commitmentState struct {
 	// cfg is a structure which carries all dependable fields/handlers
 	// which may affect behaviour of the service.
 	cfg *ChannelLinkConfig
+
+	// upstream is a channel which responsible for propagating the
+	// received from remote peer messages, with which we have an opened
+	// channel, to handler function.
+	upstream chan lnwire.Message
+
+	// downstream is a channel which responsible for propagating
+	// the received htlc switch packet which are forwarded from anther
+	// channel to the handler function.
+	downstream chan *htlcPacket
+
+	started  int32
+	shutdown int32
+	wg       sync.WaitGroup
+	quit     chan struct{}
+}
+
+// A compile time check to ensure channelLink implements the ChannelLink
+// interface.
+var _ ChannelLink = (*channelLink)(nil)
+
+// Start starts all helper goroutines required for the operation of the
+// channel link.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Start() error {
+	if !atomic.CompareAndSwapInt32(&l.started, 0, 1) {
+		log.Warn("channel link(%v): already started", l)
+		return nil
+	}
+
+	log.Info("channel link(%v): starting", l)
+
+	l.wg.Add(1)
+	go l.htlcHandler()
+
+	return nil
+}
+
+// Stop gracefully stops all active helper goroutines, then waits until they've
+// exited.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Stop() {
+	if !atomic.CompareAndSwapInt32(&l.shutdown, 0, 1) {
+		log.Warn("channel link(%v): already stopped", l)
+		return
+	}
+
+	log.Info("channel link(%v): stopping", l)
+
+	close(l.quit)
```

Comment:
I remember commenting on this on previous review cycles, shortly, this may cause the deadlock.